### PR TITLE
Allow Boolean mask indexing for `param_shift_hessian`

### PR DESCRIPTION
**Pull request by dwierichs**
Implements the option to pass a Boolean `array_like` as `argnum` to `param_shift_hessian`, to allow, for example, for computation of the diagonal alone. This is part of the feature request #2195 that was implemented to its major extent in #2319, but Boolean indexing was left open.
Internally, any provided `argnum` is converted to such a Boolean array, so that the overall code complexity is not increased as far as I can tell.

### CODE REVIEW

**Comment by dwierichs** [comment]
@cvjjm Does this work the way you would expect?

**Comment by dime10** [comment]
Thank you for adding this @dwierichs, I'll have a look at this shortly :)

**Comment by dwierichs** [comment]
Great @cvjjm! Regarding higher-dimensional masks, note that currently the `argnum` keyword argument always refers to _tape_ parameters, and not to QNode parameters. That is, while applying `param_shift_hessian` to a QNode returns a Hessian of the expected shape (`(qnode_outputshape, parshape, parshape)`), the `argnum` argument is currently only applied at the tape level, so that it should always be 2d. This is certainly a discrepancy that would be good to remove as soon as possible, but currently there might not be time yet to do it. 
@dime10 @josh146 what do you think about the priority of allowing `argnum` to act on the level of QNode arguments?

**Comment by dime10** [comment]
> what do you think about the priority of allowing argnum to act on the level of QNode arguments?

@dwierichs This is definitely a point of confusion I've thought about before, in that `argnum` for the gradient transforms refers to *trainable* tape parameters only (that is it subsets the set of trainable tape parameters). `argnum` actually setting the trainable tape parameters would make more sense to me, either directly when the input is a tape, or indirectly by referring to QNode arguments from which the set of trainable tape parameters is then defined.

File: pennylane/gradients/parameter_shift_hessian.py [hunk 1]
```diff
@@ -120,10 +184,9 @@ def expval_hessian_param_shift(
             tapes.
     """
     # pylint: disable=too-many-arguments, too-many-statements
-    argnum = tape.trainable_params if argnum is None else argnum
     h_dim = tape.num_params
 
-    unshifted_coeffs = {}
+    c_unshifted = {}
```
Comment (dime10) [inline pennylane/gradients/parameter_shift_hessian.py:189]: Slight loss in readability here 😅

File: pennylane/gradients/parameter_shift_hessian.py [hunk 2]
```diff
@@ -135,47 +198,24 @@ def expval_hessian_param_shift(
...
-    for i in range(h_dim):
-        # The diagonal recipe is None if the parameter is not trainable or not in argnum
-        if diag_recipes[i] is None:
-            hessian_coeffs.extend([None] * (h_dim - i))
+    for i, j in it.combinations_with_replacement(range(h_dim), r=2):
+        if i == j:
+            add_unshifted, c_unshifted[(i, i)] = _generate_diag_tapes(
+                tape, i, diag_recipes, add_unshifted, hessian_tapes, hessian_coeffs
+            )
             continue
```
Comment (dime10) [inline pennylane/gradients/parameter_shift_hessian.py:206]: Does this mean the diagonal elements are not masked out (as done below with `if not argnum[i, j]`) ?

**Review by dime10 — COMMENTED** [review]
Thank you @dwierichs great work 💯

File: pennylane/gradients/parameter_shift_hessian.py [hunk 3]
```diff
@@ -36,52 +35,118 @@
...
+    return argnum
+
+
 def _collect_recipes(tape, argnum, diff_methods, diagonal_shifts, off_diagonal_shifts):
     r"""Extract second order recipes for the tape operations for the diagonal of the Hessian
     as well as the first-order derivative recipes for the off-diagonal entries.
     """
+    diag_argnum = list(qml.math.where(qml.math.diag(argnum))[0])
+    offdiag = argnum ^ qml.math.diag(qml.math.diag(argnum))
+    offdiag_argnum = list(qml.math.where(qml.math.any(offdiag, axis=0))[0])
```
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/parameter_shift_hessian.py:63]: Ah, this is indeed a bit mean. Some hopefully useful remarks:
1. For the diagonal, we need second-order shift rules in a single parameter. Therefore we store in `diag_argnum` which of those shift rules we will need because `argnum[i, i]=True`.
2. For the off-diagonal entries, we will require first-order shift rules, that are then later combined into two-parameter shift rules. However, we only want to obtain and process each of these first-order rules only once, and we need the `i`-th parameter's rule only if _any_ of the entries in the `i`-th column (or row, the argnum mask is symmetric) is set to `True`. To achieve this, we first subtract the diagonal (Read `^` as `-`, which is not implemented for Boolean arrays), and then find the indices for which _any_ of the rows (`axis=0` in `qml.math.any`) contains a `True` value.

tl;dr: 
>For the off-diagonal argnum, it is still 1D but now contains the indices of the mask entries which where true in either of the 2 dimensions?

True, if you replace "in either of the 2 dimensions" by "in at least one row".

The conversion to index-based argnum could be avoided, making both `diag_argnum` and `offdiag_argnum` 1D Boolean arrays, if it is very annoying to read.

File: pennylane/gradients/parameter_shift_hessian.py [hunk 4]
```diff
@@ -36,52 +35,118 @@
...
+
     diag_recipes = []
     partial_offdiag_recipes = []
     for i in range(tape.num_params):
-        if i not in argnum or diff_methods[i] == "0":
+        if not i in diag_argnum or diff_methods[i] == "0":
```
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/parameter_shift_hessian.py:68]: Yes, after answering the above and thinking about it again, probably that's just cleaner. I'll give it a shot!

File: pennylane/gradients/parameter_shift_hessian.py (hunk 2 above)
Comment (dwierichs) [reply pennylane/gradients/parameter_shift_hessian.py:206]: Yes they are but within `_generate_diag_tapes` :facepalm: Will definitely move that out...

File: pennylane/gradients/parameter_shift_hessian.py [hunk 5]
```diff
@@ -241,10 +282,12 @@ def param_shift_hessian(tape, argnum=None, diagonal_shifts=None, off_diagonal_sh
 
     Args:
         tape (pennylane.QNode or .QuantumTape): quantum tape or QNode to differentiate
-        argnum (int or list[int] or None): Parameter indices to differentiate
+        argnum (int or list[int] or array_like or None): Parameter indices to differentiate
             with respect to. If not provided, the Hessian with respect to all
             trainable indices is returned. Note that the indices refer to tape
             parameters both if ``tape`` is a tape, and if it is a QNode.
+            If an ``array_like`` is provided, it is expected to be a two-dimensional
+            Boolean mask.
```
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/parameter_shift_hessian.py]: Yes, that is a very good hint. Not loving the system of how we are using `argnum` overall here, as you also noted already. But that's for another PR...

File: tests/gradients/test_parameter_shift_hessian.py [hunk 6]
```diff
@@ -899,6 +1023,34 @@ def circuit(x, y):
...
+        dev = qml.device("default.qubit", wires=2)
+
+        @qml.qnode(dev, max_diff=2, diff_method="parameter-shift")
+        def circuit(par):
+            qml.RX(par[0], wires=0)
+            qml.CRY(par[1], wires=[0, 1])
+            qml.CNOT(wires=[0, 1])
+            qml.CRY(par[2], wires=[0, 1])
+            return qml.probs(wires=[0, 1])
+
```
[the remark this answers is not in the document]
Comment (dwierichs) [reply tests/gradients/test_parameter_shift_hessian.py:1045]: Good call, will implement a check and a corresponding test!

**Review by dime10 — APPROVED** [review]
Thanks @dwierichs, looks good 🎉
